src/vsr.zig

```diff
@@ -1098,3 +1102,219 @@ test "ViewChangeHeaders.view_for_op" {
     try std.testing.expect(std.meta.eql(headers.view_for_op(5, 12), .{ .min = 0, .max = 7 }));
     try std.testing.expect(std.meta.eql(headers.view_for_op(0, 12), .{ .min = 0, .max = 7 }));
 }
+
+/// The headers of a SV or DVC message.
+const ViewChangeHeadersArray = struct {
+    array: Headers.Array,
+
+    pub fn root(cluster: u32) ViewChangeHeadersArray {
+        var array = Headers.Array{ .buffer = undefined };
+        array.appendAssumeCapacity(Header.root_prepare(cluster));
+        return ViewChangeHeadersArray.init(array);
+    }
+
+    fn init(array: Headers.Array) ViewChangeHeadersArray {
+        Headers.ViewChangeSlice.verify(array.constSlice());
+        return .{ .array = array };
+    }
+
+    pub fn build(
+        options: struct {
+            op_checkpoint: u64,
+            /// The last view_change_headers_max headers of the journal, starting with the head op
+            /// then descending, skipping over all gaps.
+            current: struct {
+                headers: Headers.Array,
+                view: u32,
+                log_view: u32,
+                log_view_primary: bool,
+            },
+            // The vsr_headers from the working superblock.
+            durable: struct {
+                headers: Headers.ViewChangeSlice,
+                view: u32,
+                log_view: u32,
+                log_view_primary: bool,
+            },
+        },
+    ) ViewChangeHeadersArray {
+        const current = options.current;
```

Comment:
large memcpy?

Reply:
Yeah, I'm thinking I'll change these from `BoundedArray`s to `*BoundedArray`s.
Then the physical allocation will be part of the replica, e.g.:

```
        view_change_headers: vsr.Headers.ViewChangeArray = .{ .buffer = undefined },
        journal_headers: vsr.Headers.Array = .{ .buffer = undefined },
```

which is a little awkward (they aren't really state, just scratch space for `create_view_change_headers()` to put things).

Reply:
Ah, I think, because it is const, zig should be smart enough to treat this as an alias, right?

Reply:
I think it should be, but isn't. In this particular case it could be ` = &options.current`, though.

Reply:
But that still involves a memcopy when the function is called.

Reply:
https://github.com/tigerbeetledb/tigerbeetle/pull/447/commits/2e8be5452d1fa4cdbf923547d83fc5f27350fb38